### PR TITLE
Add expected result tests

### PR DESCRIPTION
**Pull request by eellison**
With upcoming PRs like https://github.com/pytorch/vision/pull/1333 which make a number of modifications to the code, its important that we do not silently change model results.

This sets up expected results for the tests in `test_models.py`. It has the same usage as in pytorch main repo, where you add `--accept` on the cmdline to accept a new result. 

For some of the results, the pickle file is quite large, and we could consider committing summaries of the tensor results instead of the full thing.

### CODE REVIEW

File: test/test_models.py [hunk 1]
```diff
@@ -88,6 +98,7 @@ def _test_detection_model(self, name):
         out = model(model_input)
         self.assertIs(model_input[0], x)
         self.assertEqual(len(out), 1)
+        self.assertExpected(out, name)
```
Comment (eellison) [inline test/test_models.py]: ` maskrcnn_resnet50_fpn_expect.pkl ` is kind of large here - `+34.3 MB` maybe we could check something like sum and variance here instead of the actual tensor value ?

Comment (fmassa) [reply test/test_models.py]: Yeah, 34MB is a bit large. One thing that @fbbradheintz is doing in https://github.com/pytorch/vision/pull/1373 is to select a subset of the output elements. He has done this for classification models for now, not sure what his thoughts are about detection models

Comment (eellison) [reply test/test_models.py]: I started with detection models sort of arbitrarily, mostly just in anticipation of https://github.com/pytorch/vision/pull/1333. I don't see any reason why the same approach shoudn't apply to other types of models.

Agreed 34MB is too large. Current ideas are:
- moments of the results. we can also pass in a custom tolerance to `assertExpected` if we think it will be too fragile.
- subset of the output elements. One suggestion to do this is by flattening all tensor outputs, and then taking every random `nth` index such that the resulting tensor isn't more than `X` number of elements (100?).
- both ? this would be more robust to sparse tensors  
 
 
@fbbradheintz what do you think?

**Comment by eellison** [comment]
Looks like it's just `test_maskrcnn_resnet50_fpn` which is unstable right in this test

File: test/test_models.py [hunk 2]
```diff
@@ -87,6 +97,33 @@ def _test_detection_model(self, name):
...
+            return flat_tensor[ith_index - 1::ith_index]
+
+        def compute_mean_std(tensor):
+            # can't compute mean of integral tensor
+            tensor = tensor.to(torch.double)
+            mean = torch.mean(tensor)
+            std = torch.std(tensor)
+            return {"mean": mean, "std": std}
+
+        # maskrcnn_resnet_50_fpn numerically unstable across platforms, so for now
```
Comment (fmassa) [inline test/test_models.py]: Hum, interesting. What kind of unstable behavior are you observing?

File: test/common_utils.py [hunk 3]
```diff
@@ -14,3 +20,135 @@ def get_tmp_dir(src=None, **kwargs):
...
+        if subname:
+            expected_file += "_" + subname
+            subname_output = " ({})".format(subname)
+        expected_file += "_expect.pkl"
+        expected = None
+
+        def accept_output(update_type):
+            print("Accepting {} for {}{}:\n\n{}".format(update_type, munged_id, subname_output, output))
+            torch.save(output, expected_file)
+            MAX_PICKLE_SIZE = 100 * 1000  # 100 KB
```
Comment (fmassa) [inline test/common_utils.py]: Cool, I think 100KB is ok as an upper limit, although twice as much as the largest (text) file we currently have in torchvision.

File: test/common_utils.py [hunk 4]
```diff
@@ -14,3 +20,135 @@ def get_tmp_dir(src=None, **kwargs):
...
+        else:
+            return object
+
+
+def map_nested_tensor_object(object, tensor_map_fn):
+    impl = MapNestedTensorObjectImpl(tensor_map_fn)
+    return impl(object)
+
+
+# adapted from TestCase in torch/test/common_utils
```
Comment (fmassa) [inline test/common_utils.py]: could you point out the main changes / simplifications compared to what we have in `common_utils`, just for awareness? Did you just removed functionality, or did you change a few things as well?

File: test/test_models.py [hunk 5]
```diff
@@ -1,10 +1,19 @@
...
 from torchvision import models
 import unittest
 import traceback
+import random
 
+def set_rng_seed(seed):
+    torch.manual_seed(seed)
+    random.seed(seed)
+    if torch.cuda.is_available():
+        torch.cuda.set_rng_state(seed)
```
Comment (fmassa) [inline test/test_models.py]: I think this is not being called correctly, as `torch.cuda.set_rng_state` accepts a `state` object, which is a byte tensor.

I think we don't need to manually set the CUDA seed anymore, because it's done [by default when `torch.manual_seed` is called](https://github.com/pytorch/pytorch/blob/effcc398c447ce2aa6752354d64ef3d7ac5f004b/torch/random.py#L21-L34)

File: test/test_models.py [hunk 6]
```diff
@@ -42,7 +51,7 @@ def get_available_video_models():
 ]
 
 
-class Tester(unittest.TestCase):
+class Tester(TestCase):
```
Comment (fmassa) [inline test/test_models.py]: What do you think if we now change the name of the `Tester` to be something like `ModelTester`, so that the expect files have a common prefix?

File: test/test_models.py [hunk 7]
```diff
@@ -79,6 +88,7 @@ def _test_segmentation_model(self, name):
         self.assertEqual(tuple(out["out"].shape), (1, 50, 300, 300))
 
     def _test_detection_model(self, name):
+        set_rng_seed(0)
```
[the remark this answers is not in the document]
Comment (fmassa) [reply test/test_models.py]: Sounds good to me

**Comment by peterjc123** [comment]
@fmassa But the nightly builds are working fine. https://dev.azure.com/pytorch/PyTorch/_build/results?buildId=15002

**Comment by peterjc123** [comment]
@ezyang Could you please figure out how I can upload the docker image so that we can use VS2017?

File: test/test_models.py (hunk 1 above)
Comment (fbbradheintz) [reply test/test_models.py]: I'm a fan of the sampling method, though I may change the sampling number to be a prime number an order of magnitude or two less than the number of output elements, just to avoid the possibility of alignment artifacts.

Comment (fbbradheintz) [reply test/test_models.py]: Actually, looking over code again, that's not going to matter for one-dimensional classification outputs. I'd still be wary of alignment issues on multi-dimensional outputs.

Comment (fmassa) [reply test/test_models.py]: about multi-dimensional tensors, I'd just check that the shape of the tensors are also the same, and then we should be good to go?
